server/src/main/java/org/elasticsearch/common/util/concurrent/QueueResizingEsThreadPoolExecutor.java

```diff
@@ -151,8 +151,14 @@ protected void afterExecute(Runnable r, Throwable t) {
         final long totalNanos = totalTaskNanos.addAndGet(taskNanos);
 
         final long taskExecutionNanos = timedRunnable.getTotalExecutionNanos();
-        assert taskExecutionNanos >= 0 : "expected task to always take longer than 0 nanoseconds, got: " + taskExecutionNanos;
-        executionEWMA.addValue(taskExecutionNanos);
+        if (taskExecutionNanos > 0) {
+            executionEWMA.addValue(taskExecutionNanos);
+        } else {
+            // wrapper (e.g. ContextPreservingAbstractRunnable) threw an exception before being able to run this task
+            assert taskExecutionNanos == -1 && timedRunnable.isStarted() == false && t != null :
```

Comment:
@dakrone I'm not really sure about this, because the exception that `ContextPreservingAbstractRunnable` throws when shut down is re-thrown after we get here here, and goes uncaught, so the tests that are failing here look like they'd fail anyway. I'm also unsure whether we want to continue with the rest of the method, or bail out, or something else entirely.